Add Character Case Rules

jakejohns opened the pull request:
Adds Filter/Sanitize for: 
- `caseUpper` : `strtoupper`, 
- `caseLower` : `strtolower`, 
- `caseTitle` : `ucwords`, 
- `caseLowerFirst` : `lcfirst`, 
- `caseUpperFirst` : `ucfirst`. 

All use `mbstring` if available. 

The 'first' rules are a little odd because I think you gotta do that since there's no `mb_ucfirst` or `mb_lcfirst`.

jakejohns committed:
Add Character Case Rules

- Add CaseLower (lowercase)
- Add CaseUpper (UPPERCASE)
- Add CaseTitle (Title Case)

jakejohns committed:
Add more character case rules

- Add CaseLowerFirst (lOWERCASE first char)
- Add CaseUpperFirst (Uppercase first char)
- Change AbstractCharCase to extends AbstractStrlen

harikt commented on src/Rule/AbstractCharCase.php:
```diff
@@ -0,0 +1,118 @@
+ * @package Aura.Filter
+ *
+ */
+abstract class AbstractCharCase extends AbstractStrlen
```
may be you missed to add this class? I am looking where the `$this->substr` is called over https://github.com/auraphp/Aura.Filter/pull/106/files#diff-b24d614b81004a326954313c020d1a83R93

jakejohns replied:
I think you're asking where the `substr` method resides?

`AbstractStrlen` is preexisting, and contains this method. It's used for the previously existing 'string length' functions. 

The 2nd commit, which adds `ufcirst` and `lcfirst` functions, changes the `AbstractCharCase` class to extend `AbstractStrlen`, since there's not `mbstring` functions which can do uc/lcfirst.

Perhaps there's a more clear solution? Replicate the code? Traits? Different naming convention?

harikt replied:
Thank you. I missed to see the class some how. Sorry about it.

harikt commented on src/Locator/ValidateLocator.php:
```diff
@@ -37,6 +37,11 @@ protected function initFactories(array $factories)
             'between'               => function () { return new Validate\Between(); },
             'bool'                  => function () { return new Validate\Boolean(); },
             'callback'              => function () { return new Validate\Callback(); },
+            'caseLower'             => function () { return new Validate\CaseLower(); },
```
I don't think validator probably need to check if it is lower / upper. Probably users can pass it to sanitize than in validation. But I don't have trouble though. Not sure if there is a right use case.

harikt commented:
one thing probably I feel missing is how we can have white spaces to be allowed in characters. 

Eg: `name` => `Hari KT` . It is `alpha` , but with spaces. It may be good to have I guess. Thoughts ? or am I missing to see it already implemented.

We currently have regex though.

jakejohns commented:
I do believe these expressly handle white space. Specifically, this is why the validation does not use `ctype_upper`, etc. I figure if one wants to validate for all alpha, one would use that validation as well.

That being said, the validation comparing the equivalency of the raw and
converted string may be unnecessary overhead. There might be better regex to
accomplish this, but I'm not totally clear on how to implement that and support
non-`[A-Z]` characters, etc.

harikt commented:
Thank you @jakejohns for the PR.

pmjones commented:
Hi @jakejohns, I wanted to get a stable release out before addressing these.

My only comment at this point is that the names don't "read" as well as I think they might. For example, `$filter->validate('field')->is('caseUpper')` does not read as well as `$filter->validate('field')->is('upperCase')`. You get what I mean there?

jakejohns commented:
Right on. I did that so they'd group together in alpha order, like strlen.
I assume you want me to rename the actual classes as well as the names in the locator. 
- caseUpper -> uppercase
- caseLower -> lowercase
- caseTitle -> titlecase
- caseLowerFirst -> lowercaseFirst?
- caseUpperFirst -> uppecaseFirst?

pmjones commented:
I think those look just fine.

harikt commented:
@pmjones isn't the case `C` was upper ?

EDIT : may be not, for it is a single letter word.

pmjones commented:
@jakejohns I'm going to roll another Filter release today, we can get your stuff into the next one after we settle on the names. At your convenience, of course; no hurry. :-)

jakejohns committed:
Rename Char Case rules as per #106

jakejohns committed:
Add single/blank support for char case rules

jakejohns commented:
My apologies, travel and work intervene. 
I've renamed the in what I believe was the intended fashion.
I also found that there was an issue with the "first" rules not handling single characters correctly, so I addressed that as well, I think.

Let me know. Thanks.

pmjones commented:
Dude, no need to apologize, we're all on volunteer time here. :-)

pmjones commented:
Merged. I may want to tweak a bit before next release; will discuss with you as needed. Thanks!
